lib/e3db/client.rb

```diff
@@ -559,24 +621,226 @@ def revoke(type, reader_id)
     end
 
+    # Gets a list of record types that others have shared with this
+    # client.
+    #
+    # @return [Array<IncomingSharingPolicy>]
     def incoming_sharing
       url = get_url('v1', 'storage', 'policy', 'incoming')
       resp = @conn.get(url)
       json = JSON.parse(resp.body, symbolize_names: true)
       return json.map {|x| IncomingSharingPolicy.new(x)}
     end
 
+    # Create and return an (encrypted) secret key for encrypting the
+    # given record type. Can be saved for use with {#encrypt_record},
+    # {#encrypt_new_record} and {#decrypt_record} later.
+    #
+    # @return [EAK]
+    def create_writer_key(type)
+      id = @config.client_id
+      begin
+        put_access_key(id, id, id, type, new_access_key)
+      rescue Faraday::ClientError => e
+        # Ignore 403, as it means a key already exists. Otherwise, raise.
+        if e.response[:status] != 403
+          raise e
+        end
+      end
+
+      get_eak(id, id, type)
+    end
+
+    # Retrieve an (encrypted) key for reading records written by the
+    # given writer for the given user with the given type (assuming
+    # the writer previously established sharing with this client).
+    #
+    # @return [EAK]
+    def get_reader_key(writer_id, user_id, type)
+      get_eak(writer_id, user_id, type)
+    end
+
+    # Encrypts a record, using the encrypted key provided.
+    #
+    # +plain_record+ should be a Record instance to encrypt. +eak+ should
+    # be an encrypted secret key, obtained via {#create_writer_key}.
+    #
+    # @return [Record] An instace containg the encrypted data.
+    def encrypt_record(plain_record, eak)
+      cache_key = [plain_record.meta.writer_id, plain_record.meta.user_id, plain_record.meta.type]
+      if ! @ak_cache.has_key? cache_key
+        @ak_cache[cache_key] = { :eak => eak, :ak => decrypt_box(eak.eak, eak.authorizer_public_key.curve25519, @private_key) }
+      end
+      ak = @ak_cache[cache_key][:ak]
+
+      encrypted_record = Record.new(meta: plain_record.meta, data: Hash.new)
+      plain_record.data.each do |k, v|
+        dk =   new_data_key
+        efN =  secret_box_random_nonce
+        ef =   RbNaCl::SecretBox.new(dk).encrypt(efN, v)
+
+        edkN = secret_box_random_nonce
+        edk =  RbNaCl::SecretBox.new(ak).encrypt(edkN, dk)
+
+        encrypted_record.data[k] = [edk, edkN, ef, efN].map { |f| base64encode(f) }.join(".")
+      end
+
+      encrypted_record
+    end
+
+    # Encrypt a new record consisting of the given data.
+    #
+    # +type+ is a string giving the record type. +data+ should be a
+    # dictionary of string values. +plain+ should be a dictionary of
+    # string values. +id+ should be the ID of the client creating the
+    # record. +eak+ should be an encrypted secret key, obtained via
+    # {#create_writer_key}.
+    #
+    # @return [Record] An instance containing the encrypted data.
+    def encrypt_new_record(type, data, plain, id, eak)
```

Comment:
Maybe we could rename to offer `encrypt_record(type, data, ...)` and `encrypt_existing(record, ...)`? It could offer better symmetry and steer the user toward their typical use case.